### DISCUSSION TITLE
Section 5: T-Swap - Help me better understand the mathematics around getInputAmountBasedOnOutput & getOutputAmountBasedOnInput

### DISCUSSION THREAD

**Discussion by Damyanov7** [Cyfrin/security-and-auditing-full-course-s23#178] [Q&A]
I'm trying to do so manual calculations to check if calculations are correct, but I'm most likely missing something.

In the course Patrick mentioned that 10000 should be 1000.

The line that bothers me is:

```
return
            ((inputReserves * outputAmount) * 10000) / // 10_000 =>1_000 as mentioned above
            ((outputReserves - outputAmount) * 997);
```

I'm assuming that I have 2 tokens -  A and B where:
TokenA Reserve: 100_000
TokenB Reserve: 300_000

User wants to get out 500 of TokenA, he needs to find how much TokenB he must input.

The formula with values placed in place of variables is:
```
inputReserves = 300_000
outputAmount = 500
outputReserves = 100_000

We get: ((300_000 * 500) * 1_000) / ((100_000 - 500) * 997)
```

Placing this in the calculator results to: 1512.07391017
The ratio between TokenA and TokenB is 1 / 3
`=> 1512.07391017 of TokenB will result in 1512.07391017 / 3 = 504.024636723 (TokenA)`

0.3% from 504.024636723 should result in 500 (The amount that the user wanted) and the rest remains in the protocol as reward for LPs, but in reality =>

```
0.3% == 3 / 1000
(3/1000)*504.024636723 = 1.51207391017
1.51207391017 != 4.024636723
```

Where do I go wrong?

**Comment by EngrPips** [comment]
Hello @Damyanov7, The below statements become incorrect the moment the user supplied the `1512.07391017` of token A to get the `500` token B. 

```js
The ratio between TokenA and TokenB is 1 / 3
```

The ratio changes in a pool when you supply one token to get another. The ratio between TokenA and TokenB being 1/3 is correct before the user supplied the `1512.07391017` of token A to the pool.

**Reply by Damyanov7** [reply]
Hi @EngrPips,
Yes, That I understand.
1. I don't understand what amount of tokenA will  1512.07391017 of tokenB produce?
1.1 How is it being calculated?
1.2 User initially wanted 500 of tokenA, but the actual produced equivalent for 1512.07391017 is larger, isn't the difference supposed to be the 0.3% from the produced number?
1.2.1 If yes, why when I substract 0.3% from produced tokenA number I do not get the exact 500 tokenA?

**Reply by EngrPips** [reply]
So, This calculation is clearer than it appears to be. 

**POOL STATE  1**
Token A reserve = 100_000
Token B reserve = 300_000
TOKEN A / TOKEN B = 100_000/300_000

A user comes to move the pool state from `STATE 1` to `STATE 2` by taking out 500 out of the `Token A` reserve. We calculated what amount of `Token B` the user must put in the pool to be eligible to take out that 500 of  `Token A` to be `1512.07391017` [According to your calculation].

Now let's see what the state of the pool is now that the user has successfully taken 500 out of  `Token A` by depositing  1512.07391017 of `Token B.`

**POOL STATE 2**
Token A reserve = 100_000 - 500 = 99_500
Token B reserve = 300_000 + 1512.07391017 = 301_512.0739
TOKEN A / TOKEN B = 99_500/ 301_512.0739

So let's assume another user wants to swap now `1512.07391017` of Token B to get Token A you might expect them to get back 500 Token A as that was the amount a user got initially when they supplied `1512.07391017` of Token B. However, it is worth noting that the ratio of Token A to Token B changed slightly when the first swap was carried out. So we won't get exactly what we expected. This is particular to the constant product formula, and you might want to take some extra time to study more about the `constant product formula` and how it works to the finest details of it.

**Reply by Damyanov7** [reply]
I understand that after swapping the ration between the 2 tokens change.

I'm still confused about the fee and from which token it is taken.

In the example we've discussed, the pool state after the swap is as follows and as you said:
POOL STATE 2
Token A reserve = 100_000 - 500 = 99_500
Token B reserve = 300_000 + 1512.07391017 = 301_512.0739
TOKEN A / TOKEN B = 99_500/ 301_512.0739

Here we can see that user supplied  `1512.07391017` tokenB `input`.
And he got 500 tokenA `output`.

But the T-Swap README states:

```
LP Example
LP A adds 1,000 WETH & 1,000 USDC to the USDC/WETH pool
They gain 1,000 LP tokens
LP B adds 500 WETH & 500 USDC to the USDC/WETH pool
They gain 500 LP tokens
There are now 1,500 WETH & 1,500 USDC in the pool
User A swaps 100 USDC -> 100 WETH.
The pool takes 0.3%, aka 0.3 USDC.
The pool balance is now 1,400.3 WETH & 1,600 USDC
aka: They send the pool 100 USDC, and the pool sends them 99.7 WETH
```
Lets assume that WETH == tokenA, USDC == tokenB
As described by documentation above, user sends 100 tokenB (USDC) / `Input`
And gets back 99.7 tokenA (WETH) / `output`

So in the case you mentioned, the fee is taken from the input.
In the example of the README in the T-Swap docs, the fee is taken from the output.

Why is that?

**Reply by EngrPips** [reply]
That's an interesting question. You know there is no restriction on which token you can take the fee from. That varies from protocol to protocol. What is essential is to take the fee and ensure the accounting is accurate. I just gave an example to make us understand. It is not the industry's way of doing things in any way.

**Reply by Damyanov7** [reply]
I'll leave the discussion open in case someone also has something to say about it, `but in conclusion -> From which token to take fee and how to take it is protocol specific / depending on the design decision.`. Please react to this if you agree if I understood right.

**Reply by EngrPips** [reply]
Thats accurate.

**Reply by PatrickAlphaC** [reply]
You got it @Damyanov7 !
